storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java

```diff
@@ -0,0 +1,776 @@
+        dynamicState.container.kill();
+        Future<Void> pendingDownload = null;
+        if (dynamicState.newAssignment != null) {
+            pendingDownload = staticState.localizer.requestDownloadBaseTopologyBlobs(dynamicState.newAssignment, staticState.port);
+        }
+        Time.sleep(staticState.killSleepMs);
+        return dynamicState.withPendingLocalization(dynamicState.newAssignment, pendingDownload).withState(MachineState.KILL);
+    }
+    
+    /**
+     * Kill the current container and relaunch it.  (Something odd happened)
+     * PRECONDITION: container != null
+     * @param dynamicState current state
+     * @param staticState static data
+     * @return the next state
+     * @throws Exception 
+     */
+    static DynamicState killAndRelaunchContainer(DynamicState dynamicState, StaticState staticState) throws Exception {
+        assert(dynamicState.container != null);
+        
+        dynamicState.container.kill();
+        Time.sleep(staticState.killSleepMs);
+        
+        //any stop profile actions that hadn't timed out yet, we should restart after the worker is running again.
+        HashSet<TopoProfileAction> mod = new HashSet<>(dynamicState.profileActions);
+        mod.addAll(dynamicState.pendingStopProfileActions);
+        return dynamicState.withState(MachineState.KILL_AND_RELAUNCH).withProfileActions(mod, Collections.<TopoProfileAction> emptySet());
+    }
+    
+    /**
+     * Clean up a container
+     * PRECONDITION: All of the processes have died.
+     * @param dynamicState current state
+     * @param staticState static data
+     * @param nextState the next MachineState to go to.
+     * @return the next state.
+     */
+    static DynamicState cleanupCurrentContainer(DynamicState dynamicState, StaticState staticState, MachineState nextState) throws Exception {
+        assert(dynamicState.container != null);
+        assert(dynamicState.currentAssignment != null);
+        
+        dynamicState.container.cleanUp();
+        staticState.localizer.releaseSlotFor(dynamicState.currentAssignment, staticState.port);
+        DynamicState ret = dynamicState.withCurrentAssignment(null, null);
+        if (nextState != null) {
+            ret = ret.withState(nextState);
+        }
+        return ret;
+    }
+    
+    /**
+     * State Transitions for WAITING_FOR_BLOB_LOCALIZATION state.
+     * PRECONDITION: neither pendingLocalization nor pendingDownload is null.
+     * PRECONDITION: The slot should be empty
+     * @param dynamicState current state
+     * @param staticState static data
+     * @return the next state
+     * @throws Exception on any error
+     */
+    static DynamicState handleWaitingForBlobLocalization(DynamicState dynamicState, StaticState staticState) throws Exception {
+        assert(dynamicState.pendingLocalization != null);
+        assert(dynamicState.pendingDownload != null);
+        assert(dynamicState.container == null);
```

Comment:
But the state is WaitingForBlobLocalization.  That is how this function is called.